### PR TITLE
LMS Multi-database configuration

### PR DESCRIPTION
**Pull request by brendanshean**
## WHAT
Add multiple database configuration for the LMS.

Also, identify GET requests that make writes and wrap those in an around_action that forces using the primary db as described [here](https://medium.com/grailed-engineering/distributing-database-reads-across-replicas-with-rails-6-and-activerecord-23a24aa90c84). 

## WHY
Utilizing a replica database for reads will improve site performance without increasing cost.

## HOW
Much of the configuration is adapted from an earlier [PR](https://github.com/empirical-org/Empirical-Core/pull/7292) for the CMS.

1) Update the environment configurations with something of the form:
```
config.active_record.database_selector = { delay: 0.seconds }
config.active_record.database_resolver = ActiveRecord::Middleware::DatabaseSelector::Resolver
config.active_record.database_resolver_context = ActiveRecord::Middleware::DatabaseSelector::Resolver::Session
```
2) Configure database.yml
3) Add  `connects_to database: { writing: :primary, reading: :replica }` to ApplicationRecord
4) Manual smoke test on staging that the correct database is accessed based on the type of req
uest.  GET requests (except those that do writes) should go to the replica and all others should go to primary.  
Since heroku logs only broad samples of database usage, for smoke testing I had to temporarily add logging statements to application_controller and force_db_writer_role Screen Shot 2022-08-10 at 11 19 20 AM which was then confirmed on staging:
Screen Shot 2022-08-10 at 11 03 22 AM

5) A nuance to the previous smoke test concerns GET requests that write to the database.   For now, we'll wrap them in around_action hook.   In a later PR we can individually refactor them and make them idempotent.  A listing of these offenders plus possible dead routes is [here](https://docs.google.com/spreadsheets/d/1J4xHwyBlEgHrpNP5hUa1MsZ3AEgWMsi_OrK0CgcF1oU/edit#gid=0)

| controller#action | write method |
| ----------------- | ------------- |
| activity_sessions#activity_session_from_classroom_unit_and_activity | def self.find_or_create_started_activity_session(student_id, classroom_unit_id, activity_id) |
| auth/clever#clever | update_current_user_email; and signup main |
| auth/google#online_access_callback | various before_action |
| auth/google#online_access_callback | various before_action |
| blog_posts#show | @blog_post.increment_read_count |
| cms/blog_posts#destroy | destroy |
| cms/blog_posts#unpublish | update |
| cms/users#edit | log_user_action |
| cms/users#index | log_non_user_action |
| cms/users#show | log_user_action |
| cms/users#sign_in | log_user_action |
| coteacher_classroom_invitations#accept_pending_coteacher_invitations | ClassroomsTeacher.create |
| coteacher_classroom_invitations#reject_pending_coteacher_invitations | coteacher_invitation.destroy |
| schools#select_school | school = School.find_or_create_by |
| statuses#database_write | Blog.first.touch |
| students#demo_ap | Demo::ReportDemoAPCreator.create_demo(nil) |
| students#join_classroom | Associators::StudentsToClassrooms.run(current_user, classroom) |
| students#student_demo | Demo::ReportDemoDestroyer.destroy_demo(nil); Demo::ReportDemoCreator.create_demo(nil) |
| teachers/classroom_manager#dashboard | UserMilestone.find_or_create_by(user_id: current_user.id, milestone_id: | welcome_milestone.id) |
| teachers/classroom_units#launch_lesson | cuas = ClassroomUnitActivityState.find_or_create_by( |
| teachers/classrooms#hide | classroom.save |
| teachers/progress_reports#admin_demo | set_admin_user |
| teachers/progress_reports#coach_demo | set_ap_user |
| teachers/progress_reports#demo | set_user |
| teachers/progress_reports#staff_demo | set_staff_user |
| teachers#add_school | if school_user.save |

Misc links:
  https://blog.saeloun.com/2022/06/22/rails-allows-overriding-reading-request.html
  https://stackoverflow.com/questions/50956650/multiple-posgres-databases-in-circleci-2-0
  https://stackoverflow.com/questions/62056781/adding-additional-databases-in-postgresql-using-circleci
  https://www.citusdata.com/blog/2019/05/23/rails-6-multiple-databases/
  https://medium.com/williamdesk/implement-read-replica-database-on-rails-6-101-7091d535154a

### Screenshots
(If applicable. Also, please censor any sensitive data)

### Notion Card Links
(Please provide links to any relevant Notion card(s) relevant to this PR.)

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES 
Have you deployed to Staging? | YES
Self-Review: Have you done an initial self-review of the code below on Github? | YES
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -10,11 +10,17 @@ jobs:
           PG_USER: ubuntu
           RAILS_ENV: test
           RACK_ENV: test
-      - image: postgres:10.5-alpine
+      - image: postgres:13.7-alpine
```
Comment (dandrabik) [inline .circleci/config.yml:13]: Thanks for updating these.

File: services/QuillLMS/app/controllers/cms/blog_posts_controller.rb [hunk 2]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 class Cms::BlogPostsController < Cms::CmsController
+  around_action :force_writer_db_role, only: [:destroy, :unpublish]
```
Comment (dandrabik) [inline services/QuillLMS/app/controllers/cms/blog_posts_controller.rb:4]: I think this audit will be helpful in flagging actions that use the wrong http verb like these. destroy as a `'GET'` can be dangerous

File: services/QuillLMS/spec/requests/multi_db_request_spec.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+class WidgetsController < ApplicationController
```
Comment (dandrabik) [inline services/QuillLMS/spec/requests/multi_db_request_spec.rb:5]: Nice tests here!

**Review by dandrabik — APPROVED** [review]
Really nice work on this. I like how you were able to keep this initial diff pretty small for an application-wide change. This approach gives us a nice `TODO` list of controller actions to possibly refactor.

Thanks for your work on this.

File: .circleci/config.yml (hunk 1 above)
Comment (anathomical) [inline .circleci/config.yml:13]: Good update.  Should we put a `name` on the leader the way we do on the replica?  Just for clarity?

File: services/QuillCMS/config/database.yml [hunk 4]
```diff
@@ -30,27 +30,27 @@ test:
     replica: true
 
 elastic_beanstalk_primary_env: &elastic_beanstalk_primary_env
```
Comment (anathomical) [inline services/QuillCMS/config/database.yml]: Not something to fix now, but we're no longer using Elastic Beanstalk for the CMS.  We should strip out all the old references to it at some point.

File: services/QuillLMS/app/controllers/auth/clever_controller.rb [hunk 5]
```diff
@@ -2,6 +2,8 @@
 
 module Auth
   class CleverController < ApplicationController
+    around_action :force_writer_db_role, only: :clever
```
Comment (anathomical) [inline services/QuillLMS/app/controllers/auth/clever_controller.rb]: I like wrapping the `only` value in an array even if it only has one item in it.  You do this in `ActivitySessionsController`, so might as well be consistent.

File: services/QuillLMS/app/controllers/blog_posts_controller.rb [hunk 6]
```diff
@@ -1,13 +1,15 @@
 # frozen_string_literal: true
 
 class BlogPostsController < ApplicationController
+  around_action :force_writer_db_role, only: :show
```
Comment (anathomical) [inline services/QuillLMS/app/controllers/blog_posts_controller.rb]: Again, I like arrays even for a single item.  (Or, for consistency, you can remove the array for `ActivitySessionsController`.  Whatever you prefer.)

File: services/QuillLMS/app/controllers/schools_controller.rb [hunk 7]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 class SchoolsController < ApplicationController
+  around_action :force_writer_db_role, only: :select_school
```
Comment (anathomical) [inline services/QuillLMS/app/controllers/schools_controller.rb]: Again a single item that's not an array.  Which, again, is fine if that's the pattern we prefer.

File: services/QuillLMS/app/controllers/statuses_controller.rb [hunk 8]
```diff
@@ -3,6 +3,8 @@
 class StatusesController < ApplicationController
   protect_from_forgery with: :null_session
 
+  around_action :force_writer_db_role, only: :database_write
```
Comment (anathomical) [inline services/QuillLMS/app/controllers/statuses_controller.rb]: Again a single item that's not an array.  Which, again, is fine if that's the pattern we prefer.

File: services/QuillLMS/app/controllers/teachers/classroom_manager_controller.rb [hunk 9]
```diff
@@ -4,15 +4,18 @@ class Teachers::ClassroomManagerController < ApplicationController
   include CheckboxCallback
   include CleverAuthable
   include DiagnosticReports
+  include QuillAuthentication
+  include ScorebookHelper
 
   respond_to :json, :html
+
+  around_action :force_writer_db_role, only: :dashboard
```
Comment (anathomical) [inline services/QuillLMS/app/controllers/teachers/classroom_manager_controller.rb]: Again a single item that's not an array.  Which, again, is fine if that's the pattern we prefer.

File: services/QuillLMS/app/controllers/teachers/classroom_units_controller.rb [hunk 10]
```diff
@@ -1,13 +1,17 @@
...
 
+require 'pusher'
+
 class Teachers::ClassroomUnitsController < ApplicationController
   include QuillAuthentication
-  require 'pusher'
+
   respond_to :json
+
+  around_action :force_writer_db_role, only: :launch_lesson
```
Comment (anathomical) [inline services/QuillLMS/app/controllers/teachers/classroom_units_controller.rb]: Again a single item that's not an array.  Which, again, is fine if that's the pattern we prefer.

File: services/QuillLMS/app/controllers/teachers/classrooms_controller.rb [hunk 11]
```diff
@@ -4,7 +4,11 @@ class Teachers::ClassroomsController < ApplicationController
   include CleverAuthable
 
   respond_to :json, :html, :pdf
+
+  around_action :force_writer_db_role, only: :hide
```
Comment (anathomical) [inline services/QuillLMS/app/controllers/teachers/classrooms_controller.rb]: Again a single item that's not an array.  Which, again, is fine if that's the pattern we prefer.

File: services/QuillLMS/app/controllers/teachers_controller.rb [hunk 12]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 class TeachersController < ApplicationController
+  around_action :force_writer_db_role, only: :add_school
```
Comment (anathomical) [inline services/QuillLMS/app/controllers/teachers_controller.rb]: Again a single item that's not an array.  Which, again, is fine if that's the pattern we prefer.

File: services/QuillLMS/spec/requests/multi_db_request_spec.rb [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+end
+
+RSpec.describe WidgetsController, type: :request do
+  let(:widget) { double(:widget, id: 1) }
+
+  before { Rails.application.routes.draw { resources :widgets } }
+
+  after { Rails.application.reload_routes! }
+
+  subject { http_request}
```
Comment (anathomical) [inline services/QuillLMS/spec/requests/multi_db_request_spec.rb]: Spacing inconsistency.

File: services/QuillLMS/spec/requests/multi_db_request_spec.rb (hunk 3 above)
Comment (anathomical) [reply services/QuillLMS/spec/requests/multi_db_request_spec.rb:5]: Agreed.  This is great.

**Review by anathomical — APPROVED** [review]
This looks great to me.  Some very minor flags on style and stuff, but you can take or leave whatever you want.

**Review by happythenewsad — APPROVED** [review]
LGTM

File: .circleci/config.yml (hunk 1 above)
Comment (brendanshean) [reply .circleci/config.yml:13]: Good idea

File: services/QuillLMS/spec/requests/multi_db_request_spec.rb (hunk 13 above)
Comment (brendanshean) [reply services/QuillLMS/spec/requests/multi_db_request_spec.rb]: Good eye.

File: services/QuillLMS/app/controllers/auth/clever_controller.rb (hunk 5 above)
Comment (brendanshean) [reply services/QuillLMS/app/controllers/auth/clever_controller.rb]: Old habits die hard.  I agree on keeping it consistent.

File: .circleci/config.yml (hunk 1 above)
Comment (brendanshean) [reply .circleci/config.yml:13]: I guess circleci  doesn't [like](https://app.circleci.com/pipelines/github/empirical-org/Empirical-Core/17288/workflows/323c8a5c-6789-4950-aded-2276c55aaba3/jobs/232147) the leader name key.

**Pull request merged**